### PR TITLE
Implemented file specific logging to not log to the main logger

### LINKED ISSUE #82: check why main logs contain the log messages of the file logger
Our logging system creates a main log file for each main script and a new log filer for each processed file. So ideally the main log file contains only the log messages from the main script and everything else is put in the tracker log files. This works for the file logs, that only contain the messages for each file, but the main log currently also contains ALL log files for all processed files. 

It seems that the call to `addDefaultFileLogger` indeed creates a new file but that a call to a log function like `logDebug` writes in both main and file log. We didn't have this problem when we were using the parallelised approach with foreach so I suspect that the parallelLogger layout is not working in this case. 

If there is no other solution we would have to deactivate the main logger (unregister it) and register it again after a file is processed...

The main idea is that all file related messages are put in the file logs, while the main script should only report the progress about which files were processed as well as the errors and warnings. Currently, these are also logged with the files. I guess it would be better to put the errors and warnings in the main log file...

### PR DESCRIPTION
**Pull request by KonradUdoHannes**
closes #82

Our logging library does not seem to support logging only to specific loggers. In their documentation they also recommend to just use one log file. Consequently I implemented a solution to temporarily remove the main logger, or rather all current loggers as this seems more robust.

I implemented a custom `with_*` function with `withr` to enable temporary modification and added it to the three scripts.

While this works I think it makes error tracking a little more tedious at the moment as file specific errors won't show up in the main log files. This makes it necessary to check all the other log files individually for errors/warnings.  Once we are done with this issue I recommend to log files with errors (and possibly warnings) in the main log file with their name and the number of errors if its > 0.

### CODE REVIEW

**Comment by pmayd** [comment]
Wow you are fast so I guess I will learn some R from you :)

Yes I thought as much, I guess it worked before because the main script was in a separate process so it seems log messages are separated when there is a new process.

Yes, Overall this was my goal, too, to have all error messages in a error logs file. I do not remember if the current library supports that, can we specify to which log file we want to write or which logger should be used when calling logX? 

Anyway, yes, the current approach will look through all files and filter out warnings and error. This might not be ideal but honestly it is fast enough that we have the result immediately so in my opinion we don't have to put everything in a single file it is fine to parse it later but if there is an option would be nice

**Comment by KonradUdoHannes** [comment]
We cannot specify which logger to log to explicitly. We could make more use of `with_*` wrappers to achieve such an effect but we would need a little bit of extra work to make sure corresponding error log commands/their wrappers know about where to send the error logging.

One could also work more with the log level and make use of loggers that only log at the `error` and consequently ignore `warning` `info` and `debug` messages. That would go a little bit in the described direction, but would again cause a bit of duplicate logging. For instance if we have a `error` and a `debug` logger, then a warning-level message would only get logged in the debug logger, but both loggers would pick up error-level messages.

What I proposed with regard to essentially aggregation of error messages, would mostly help for quick manual inspection. Any script that systematically scans the logs will not care given our log file sizes. So the question is a bit how much convenience for manuel inspection helps us in our workflows and deployment. If it won't help too much time is better spent working on a comprehensive error report.

**Comment by pmayd** [comment]
Yes thanks Konrad, I just thought about why not use different loggers with different logLevels so we really have a single error file BUT the main reason against this is that we don't have enough information to link this back to the file that created this problem! This is exactly why I did it the way it is now: all errors and warnings are in the file where they occurred so it is very easy to identify the source file.

So yes let's keep it functional and pragmatic here and let's concentrate on the error report rather than on other logger libraries to achieve this. We have a very good logging approach so far that works well with clean separation between files and we can work with the small error logs in real-time so to speak

File: R/logger.R [hunk 1]
```diff
@@ -29,12 +29,38 @@ setup_logger <- function(output_dir, log_name) {
...
+#'
+#' Temporarily change logging to exclusively log to specific log file
+#'
+#' @param new New filename for the logfile
+#' @param code Code to be executed
+#' @param output_root Root output folder from where to construct the log path
+
+#' @export
+with_file_logger <- withr::with_(
+    setup_file_logger,
```
Comment (pmayd) [inline R/logger.R:59]: As `withr` is a very special package and at least I was not familiar with it, it is always a good idea to explicitly state the parameter names, in this case `set=` and `reset=`

File: R/logger.R [hunk 2]
```diff
@@ -29,12 +29,38 @@ setup_logger <- function(output_dir, log_name) {
...
 #'
-#' @param output_root Output root directory for the current process.
 #' @param logfile The name of the log file.
-setup_file_logger <- function(output_root, logfile) {
+#' @param output_root Output root directory for the current process.
+#'
+#' @return returns the loggers that where previously set for usage with with_
+setup_file_logger <- function(logfile, output_root) {
+    loggers <- getLoggers()
+    clearLoggers()
```
Comment (pmayd) [inline R/logger.R:38]: why do we have to clear loggers again here if that is done already by the wrapper function `with_file_logger` before the `setup_file_logger` function is called?

**Review by pmayd — APPROVED** [review]
Cool that such a thing is possible and in the end quite elegant as it requires little changes in the code BUT it is so hard to wrap your head around if you see this for the first time...might be nice to describe the behavior a little more in the comments of the wrapper function. So basically this overwrites the original function but only when executing code with the new wrapper function. Then it executes the setup_file_logger, which deletes all previous loggers so it is exclusively logged to the new log file, but returns all old loggers so that after the code block the reset function can take over and re-register the old loggers.

**Pull request merged**